pmlc/conversion/linalg_to_pxa/pooling.cc

```diff
@@ -0,0 +1,142 @@
+  }
+  builder.create<linalg::YieldOp>(builder.getUnknownLoc(), result);
+}
+
+template <typename PoolingOpType>
+struct GeneralizePoolingOp : public OpRewritePattern<PoolingOpType> {
+  using OpRewritePattern<PoolingOpType>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(PoolingOpType op,
+                                PatternRewriter &rewriter) const override {
+    AffineMap shapesMap = op.getLoopsToShapesMap();
+
+    Type origType = op.output().getType();
+    ShapedType outputType = origType.cast<ShapedType>();
+    unsigned numIdxs = shapesMap.getNumDims();
+    auto exprs = shapesMap.getResults();
+    MLIRContext *context = op.getContext();
+    // Input's affine map
+    Type inputType = op.input().getType();
+    unsigned numInputDims = inputType.cast<ShapedType>().getRank();
+    SmallVector<AffineExpr, 4> inputExprs(exprs.begin(),
+                                          exprs.begin() + numInputDims);
+    AffineMap inputMap = AffineMap::get(numIdxs, 0, inputExprs, context);
+    // Input's affine map
+    Type windowType = op.windowDims().getType();
+    unsigned numWindowDims = windowType.cast<ShapedType>().getRank();
+    SmallVector<AffineExpr, 4> windowExprs(exprs.begin() + numInputDims,
+                                           exprs.begin() +
+                                               (numInputDims + numWindowDims));
+    AffineMap windowMap = AffineMap::get(numIdxs, 0, windowExprs, context);
+    // Output's affine map
+    SmallVector<AffineExpr, 4> outputExprs(
+        exprs.begin() + (numInputDims + numWindowDims), exprs.end());
+    AffineMap outputMap = AffineMap::get(numIdxs, 0, outputExprs, context);
+
+    // Use different body builder for different pooling ops
+    GenericOpBodyBuilder bodyBuilder;
+    if (isa<PoolingMaxOp>(op)) {
+      bodyBuilder = buildPoolingMaxOpBody;
+    } else if (isa<PoolingMinOp>(op)) {
+      bodyBuilder = buildPoolingMinOpBody;
+    } else if (isa<PoolingSumOp>(op)) {
+      bodyBuilder = buildPoolingSumOpBody;
+    } else {
+      op.emitError("Invalid pooling op.");
+    }
+
+    SmallVector<StringRef, 4> iterTypes(
+        op.iterator_types().template getAsValueRange<StringAttr>());
+    auto genericOp = createGenericOp(
+        /*builder=*/rewriter,
+        /*locationOp=*/op,
+        /*outputTypes=*/TypeRange{outputType},
+        /*inputs=*/ValueRange{op.input(), op.windowDims()},
+        /*outputs=*/ValueRange{op.output()},
+        /*numIdxs=*/numIdxs,
+        /*maps=*/ArrayRef<AffineMap>{inputMap, windowMap, outputMap},
+        /*iterTypes*/ iterTypes,
+        /*bodyBuilder=*/bodyBuilder);
+
+    // Do not replace the use of the output in the new generic op
+    op.output().replaceUsesWithIf(genericOp.getResult(0), [&](OpOperand &use) {
+      return genericOp.getOperation() != use.getOwner();
+    });
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+void populateLinalgPoolingOpGeneralizationPatterns(
```

Comment:
This is a great approach (generalize all we can and then lower generic ops to PXA).  That said, I would imagine we could probably upstream these patterns to be part of the upstream linalg generalization passes.

Reply:
Will do when I have time. There are a lot of changes.